test/pylib: ManagerClient helper to wait for server to see other servers after start/restart

alecco opened the pull request:
When starting/restarting a server, provide a way to wait for the server to see other servers.
    
Also leave the implementation methods available for manual use and update previous tests.

Fixes [#13147](https://github.com/scylladb/scylladb/issues/13147)

alecco commented:
Note: this pr consists of only 1 commit and it is based on top of existing https://github.com/scylladb/scylladb/pull/12699

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/472/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/ec4504fa8bde2eb1809465195c1daad78c87eee0:

- rebased to latest https://github.com/scylladb/scylladb/pull/12699 and removed unnecessary import

Changes in https://github.com/scylladb/scylladb/commit/32de1bab09e50a0a2823dbd88423d04202a3c07c:

- type hint for `Managerclient.driver_connect()` (Kamil)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/488/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/5f7f2f0c6882c44eab1520e46f9a5ae9bcc6af0b ([cmp](https://github.com/scylladb/scylladb/compare/32de1bab09e50a0a2823dbd88423d04202a3c07c..c3fa69392aabed9e010e3422677e6e2251bc18bc)):

- 2 helpers: see minumum n other servers vs wait to see specific other server

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/494/

alecco commented:
Changes in latest:
- rebase to get commit with close driver before reconnecting

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/511/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/525/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/712f248b20f0aa24e620254f3035b4855ddaab90:

- rebase on latest https://github.com/scylladb/scylladb/pull/12699 (driver workaround)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/621/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/b961994a86d4da569cdcabf9a576685a11e72ab1:

- rebase

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/666/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/b3fdddc7958e72969e4d5546768a42e0f3ca00c1:

- helper to wait a server to not see another in gossiper (fix https://github.com/scylladb/scylladb/commit/b3fdddc7958e72969e4d5546768a42e0f3ca00c1)

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/9ecc452d4001c08c6bffde1753edf82e4649bfe9:

- deadline as parameter (Kamil)
- combine `wait_for_others` and `others_count` (Kamil)
- fix docstrings (Kamil)
- `server_sees_others()` take `ServerNum` and resolve `IPAddress`, handle `count == 0`
- remove test imports no longer needed

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/711/

alecco commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/711/

Unrelated java issue. Restarted build.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/715/

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -239,3 +242,31 @@ async def get_host_id(self, server_id: ServerNum) -> HostID:
             raise Exception(f"Failed to get local host id address for server {server_id}") from exc
         return HostID(host_id)
+
+    async def server_sees_others(self, server_id: ServerNum, count: int, deadline: int = 45):
```
The meaning of `deadline` in `wait_for` is a time point (and note it's a float), not a time interval, we should consistently stick to one meaning if we're using the same word.

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/1b146e7296dab3a3d9df866582c92340727b7070

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -239,3 +242,31 @@ async def get_host_id(self, server_id: ServerNum) -> HostID:
+            alive_nodes = await self.api.get_alive_endpoints(server_ip)
+            if not other_ip in alive_nodes:
+                return True
+        await wait_for(_not_sees_another_server, time() + 45, period=.1)
```
Forgot about parameterizing deadline here

alecco replied:
Fixed in https://github.com/scylladb/scylladb/commit/1b146e7296dab3a3d9df866582c92340727b7070

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/1b146e7296dab3a3d9df866582c92340727b7070:

- rename deadline to interval (Kamil)
- interval as float (Kamil)

alecco commented on test/pylib/manager_client.py:
```diff
@@ -139,17 +139,25 @@ async def server_stop_gracefully(self, server_id: ServerNum) -> None:
 
-    async def server_restart(self, server_id: ServerNum) -> None:
-        """Restart specified server"""
+    async def server_restart(self, server_id: ServerNum, wait_for_others: bool = False,
```
Fixed in https://github.com/scylladb/scylladb/commit/9ecc452d4001c08c6bffde1753edf82e4649bfe9

alecco commented on test/pylib/manager_client.py:
```diff
@@ -239,3 +247,27 @@ async def get_host_id(self, server_id: ServerNum) -> HostID:
+        await wait_for(_sees_another_server, time() + 45, period=.1)
+
+    async def server_not_sees_other_server(self, server_ip: IPAddress, other_ip: IPAddress):
+        """Wait till a server sees another specific server IP"""
```
Fixed in https://github.com/scylladb/scylladb/commit/9ecc452d4001c08c6bffde1753edf82e4649bfe9

kbr-scylla commented:
Looks like there's no CI run after the latest push
triggered manually: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/754/
